tests/test_halo_data_transformer.py

```diff
@@ -0,0 +1,452 @@
+
+
+def _get_boundaries(quantity, n_halos):
+    boundary_send_interior = True
+    boundary_recv_interior = False
+
+    send_boundaries = {}
+    recv_boundaries = {}
+    for direction in [
+        NORTH,
+        NORTHWEST,
+        WEST,
+        SOUTHWEST,
+        SOUTH,
+        SOUTHEAST,
+        EAST,
+        NORTHEAST,
+    ]:
+        send_boundaries[direction] = _boundary_utils.get_boundary_slice(
+            quantity.dims,
+            quantity.origin,
+            quantity.extent,
+            quantity.data.shape,
+            direction,
+            n_halos,
+            interior=boundary_send_interior,
+        )
+        recv_boundaries[direction] = _boundary_utils.get_boundary_slice(
+            quantity.dims,
+            quantity.origin,
+            quantity.extent,
+            quantity.data.shape,
+            direction,
+            n_halos,
+            interior=boundary_recv_interior,
+        )
+
+    return send_boundaries, recv_boundaries
+
+
+def test_data_transformer_scalar_pack_unpack(quantity, rotation, n_halos):
+    original_quantity: Quantity = copy.deepcopy(quantity)
+    data_transformer = HaloDataTransformer.get_from_numpy_module(quantity.np)
+
+    send_boundaries, recv_boundaries = _get_boundaries(quantity, n_halos)
+    N_edge_boundaries = {
+        0: (send_boundaries[NORTH], recv_boundaries[SOUTH]),
+        -1: (send_boundaries[NORTH], recv_boundaries[WEST]),
+        -2: (send_boundaries[NORTH], recv_boundaries[NORTH]),
+        -3: (send_boundaries[NORTH], recv_boundaries[EAST]),
+    }
+
+    NE_corner_boundaries = {
+        0: (send_boundaries[NORTHEAST], recv_boundaries[SOUTHEAST]),
+        -1: (send_boundaries[NORTHEAST], recv_boundaries[SOUTHWEST]),
+        -2: (send_boundaries[NORTHEAST], recv_boundaries[NORTHWEST]),
+        -3: (send_boundaries[NORTHEAST], recv_boundaries[NORTHEAST]),
+    }
+
+    specification = HaloUpdateSpec(
+        shape=quantity.data.shape,
+        strides=quantity.data.strides,
+        itemsize=quantity.data.itemsize,
+        origin=quantity.metadata.origin,
+        extent=quantity.metadata.extent,
+        dims=quantity.metadata.dims,
+        numpy_module=quantity.np,
+        dtype=quantity.metadata.dtype,
+    )
+
+    data_transformer.add_scalar_specification(
+        specification,
+        N_edge_boundaries[rotation][0],
+        rotation,
+        N_edge_boundaries[rotation][1],
+    )
+    data_transformer.add_scalar_specification(
+        specification,
+        NE_corner_boundaries[rotation][0],
+        rotation,
+        NE_corner_boundaries[rotation][1],
+    )
+
+    data_transformer.compile()
+    data_transformer.async_pack([quantity])
+    data_transformer.synchronize()
+    # Simulate data transfer
+    data_transformer.get_recv_buffer().assign_from(
+        data_transformer.get_send_buffer().array
+    )
+    data_transformer.async_unpack([quantity])
+    data_transformer.finalize()
```

Comment:
Are there alternative versions of this part of the code that should be tested, e.g. a synchronous pack/unpack, or that exceptions are raised when these happen in the wrong order?

Reply:
Unfortunately if you unpack before packing you will get bad data. Difficult to check that without introducing state in the transformer, which I'd rather not do